src/core/space_interaction_API.jl

```diff
@@ -149,8 +149,9 @@ Remove all the agents of the model.
 """
 function remove_all!(model::ABM)
     for a in allagents(model)
-        remove_agent!(a, model)
+        remove_agent_from_space!(a, model)
     end
+    empty!(agent_container(model))
```

Comment:
I don't like this change because `empty!(agent_container())` isn't the API way to remove an agent. The API way to remove agents is with `remove_agent!` which calls the from model and from space methods. Later down the line (multi-agent container) this `empty!` call may not work.

This optimization is not necessary. `remove_all` is not only called incredibly unfrequently, but its performance weight is miniscule compared to eg a single `step!` call.

Reply:
yes, clearly not used that much but I would use it in this new version of sample!. But I agree it is not right in the current form, but we can easily adapt it: we creare a `remove_all_from_model` with the `empty!` call for a `StandardABM`, and the fallback of looping over agents in other cases

Reply:
I still would argue the increase in code complexity is simply not worth it, because in Agents.jl you want a minimal API more than you want a performant API. This optimization isn't particularly impactful. I would strongly favor keeping the code simpler rather than micro optimizing this part.

Unless you can convince me of course that his optimization is really impactful. How much xx performance gain we get with using `empty!`?

Reply:
Benchmark:

```julia
using Agents, BenchmarkTools, Random

model_step!(model) = nothing

function fake_model(rng, nagents)
    model = StandardABM(NoSpaceAgent; model_step!, rng)
    for i in 1:nagents
        add_agent!(model)
    end
    model
end
```

with the optimization:

```julia
julia> @benchmark sample!(model, 10^6) setup=(model = fake_model(Xoshiro(42), 10^7)) evals=1 samples=10 seconds=10
BenchmarkTools.Trial: 5 samples with 1 evaluation.
 Range (min … max):  244.839 ms … 615.717 ms  ┊ GC (min … max):  0.00% … 50.54%
 Time  (median):     305.240 ms               ┊ GC (median):     0.00%
 Time  (mean ± σ):   348.143 ms ± 151.834 ms  ┊ GC (mean ± σ):  17.88% ± 22.60%

  █  █     ██                                                 █  
  █▁▁█▁▁▁▁▁██▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  245 ms           Histogram: frequency by time          616 ms <

 Memory estimate: 114.44 MiB, allocs estimate: 1000010.
```


without:

```julia
julia> @benchmark sample!(model, 10^6) setup=(model = fake_model(Xoshiro(42), 10^7)) evals=1 samples=10 seconds=10
BenchmarkTools.Trial: 4 samples with 1 evaluation.
 Range (min … max):  1.024 s …    1.655 s  ┊ GC (min … max):  0.00% … 37.55%
 Time  (median):     1.209 s               ┊ GC (median):    12.97%
 Time  (mean ± σ):   1.274 s ± 292.504 ms  ┊ GC (mean ± σ):  18.34% ± 18.49%

  █  █                          █                          █  
  █▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  1.02 s         Histogram: frequency by time         1.66 s <

 Memory estimate: 114.44 MiB, allocs estimate: 1000010.
```

So it is significant, I would also argue that it's not particularly complex code, just more code :D

Reply:
Didn't you say that the new version is actually _slower_ for models with space? What's the slowdown there?

Reply:
> Didn't you say that the new version is actually _slower_ for models with space? What's the slowdown there?

yes if I remembered correctly it is like 2x slower in the worst case, in others it is already faster. But not worth it in this case until we don't have a faster removal of all agents.